src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/OverrideMethodsOnComparableTypes.cs

```diff
@@ -73,12 +98,70 @@ private static void AnalyzeSymbol(INamedTypeSymbol namedTypeSymbol, INamedTypeSy
             if (namedTypeSymbol.AllInterfaces.Any(t => t.Equals(comparableType) ||
                                                       (t.ConstructedFrom?.Equals(genericComparableType) ?? false)))
             {
-                if (!(namedTypeSymbol.OverridesEquals() && namedTypeSymbol.ImplementsComparisonOperators()))
+                var overridesEquals = namedTypeSymbol.OverridesEquals();
+                string comparisonOperatorsString = GetNeededComparisonOperators(namedTypeSymbol);
+
+                if (!overridesEquals && comparisonOperatorsString.Length != 0)
+                {
+                    addDiagnostic(namedTypeSymbol.CreateDiagnostic(RuleBoth, namedTypeSymbol.Name, comparisonOperatorsString));
+                }
+                else if (!overridesEquals)
+                {
+                    addDiagnostic(namedTypeSymbol.CreateDiagnostic(RuleEquals, namedTypeSymbol.Name));
+                }
+                else if (comparisonOperatorsString.Length != 0)
+                {
+                    addDiagnostic(namedTypeSymbol.CreateDiagnostic(RuleOperator, namedTypeSymbol.Name, comparisonOperatorsString));
+                }
+            }
+        }
+
+        private static string GetNeededComparisonOperators(INamedTypeSymbol symbol)
+        {
+            bool first = true;
+            StringBuilder sb = new StringBuilder();
```

Comment:
Minor: initialize to null, add `sb = sb ?? new StringBuilder();` at the start of append, and add a null check at return.